eid_iomem_pci.c

```diff
@@ -0,0 +1,221 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * Copyright (C) 2022 Eideitcom
+ */
+
+#include <linux/module.h>
+#include <linux/pci.h>
+#include <linux/pci-p2pdma.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <linux/pfn_t.h>
+
+#define PCI_VENDOR_EIDETICOM 0x1de5
+#define PCI_DEVICE_IOMEM 0x1000
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Andrew Maier <andrew.maier@eideticom.com>");
+MODULE_DESCRIPTION("Driver for the Eideticom IOMEM PCIe device");
+
+static int max_devices = 16;
+module_param(max_devices, int, 0444);
+MODULE_PARM_DESC(max_devices, "Maximum number of eid-iomem devices");
+
+static struct class *eid_iomem_class;
+static DEFINE_IDA(eid_iomem_ida);
+static dev_t eid_iomem_devt;
+
+static struct pci_device_id eid_iomem_pci_id_table[] = {
+	{ PCI_DEVICE(PCI_VENDOR_EIDETICOM, PCI_DEVICE_IOMEM), },
+	{ 0, }
+};
+MODULE_DEVICE_TABLE(pci, eid_iomem_pci_id_table);
+
+struct eid_iomem_dev {
+	struct device dev;
+	struct pci_dev *pdev;
+	int id;
+	struct cdev cdev;
+};
+
+static struct eid_iomem_dev *to_eid_iomem(struct device *dev)
+{
+	return container_of(dev, struct eid_iomem_dev, dev);
+}
+
+static int eid_iomem_open(struct inode *inode, struct file *filp)
+{
+	struct eid_iomem_dev *p;
+
+	p = container_of(inode->i_cdev, struct eid_iomem_dev, cdev);
+	filp->private_data = p;
+	//pci_p2pdma_file_open(p->pdev, filp);
+
+	return 0;
+}
+
+static int eid_iomem_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	struct eid_iomem_dev *p = filp->private_data;
+
+	return pci_mmap_p2pmem(p->pdev, vma);
+}
+
+static const struct file_operations eid_iomem_fops = {
+	.owner = THIS_MODULE,
+	.open = eid_iomem_open,
+	.mmap = eid_iomem_mmap,
+};
+
+static void eid_iomem_release(struct device *dev)
+{
+	struct eid_iomem_dev *p = to_eid_iomem(dev);
+
+	kfree(p);
+}
+
+static struct eid_iomem_dev *eid_iomem_create(struct pci_dev *pdev)
+{
+	struct eid_iomem_dev *p;
+	int err;
+
+	p = kzalloc(sizeof(*p), GFP_KERNEL);
+	if (!p)
+		return ERR_PTR(-ENOMEM);
+
+	p->pdev = pdev;
+
+	device_initialize(&p->dev);
+	p->dev.class = eid_iomem_class;
+	p->dev.parent = &pdev->dev;
+	p->dev.release = eid_iomem_release;
+
+	p->id = ida_simple_get(&eid_iomem_ida, 0, 0, GFP_KERNEL);
+	if (p->id < 0) {
+		err = p->id;
+		goto out_free;
+	}
+
+	dev_set_name(&p->dev, "eid-iomem%d", p->id);
+	p->dev.devt = MKDEV(MAJOR(eid_iomem_devt), p->id);
+
+	cdev_init(&p->cdev, &eid_iomem_fops);
+	p->cdev.owner = THIS_MODULE;
+
+	err = cdev_device_add(&p->cdev, &p->dev);
+	if (err)
+		goto out_ida;
+
+	dev_info(&p->dev, "registered");
+
+	return p;
+
+out_ida:
+	ida_simple_remove(&eid_iomem_ida, p->id);
+out_free:
+	put_device(&p->dev);
+	return ERR_PTR(err);
+}
+
+void eid_iomem_destroy(struct eid_iomem_dev *p)
+{
+	dev_info(&p->dev, "unregistered");
+	cdev_device_del(&p->cdev, &p->dev);
+	ida_simple_remove(&eid_iomem_ida, p->id);
```

Comment:
Might be better to call this inside the release function so that the same ID isn't created before the device actually goes away.